File: modules/composites.js
// ============================================================
// MODULE: composites
// Sentinel-2 and AlphaEarth embedding image builders
// ============================================================
 
var S2_COLLECTION  = 'COPERNICUS/S2_SR_HARMONIZED';
var CS_COLLECTION  = 'GOOGLE/CLOUD_SCORE_PLUS/V1/S2_HARMONIZED';
var EMB_COLLECTION = 'GOOGLE/SATELLITE_EMBEDDING/V1/ANNUAL';

// Annual cloud-free Sentinel-2 median composite
exports.getSentinel2Annual = function(year, region) {
  var start  = ee.Date.fromYMD(year, 1, 1);
  var end    = start.advance(1, 'year');
  var s2     = ee.ImageCollection(S2_COLLECTION)
    .filter(ee.Filter.date(start, end))
    .filter(ee.Filter.bounds(region));
  var csPlus = ee.ImageCollection(CS_COLLECTION);
  return s2.linkCollection(csPlus, csPlus.first().bandNames())
    .map(function(img) {
      return img.updateMask(img.select('cs').gte(0.65));
    }).select('B.*').median().clip(region);
};

// Seasonal composite with cloud-score + fallback
exports.getSeasonalComposite = function(startDate, endDate, region) {
  var csPlus = ee.ImageCollection(CS_COLLECTION)
    .filter(ee.Filter.date(startDate, endDate))
    .filter(ee.Filter.bounds(region));
  var s2 = ee.ImageCollection(S2_COLLECTION)
    .filter(ee.Filter.date(startDate, endDate))
    .filter(ee.Filter.bounds(region));
  var withCs = s2.linkCollection(csPlus, csPlus.first().bandNames())
    .map(function(img) {
      return img.updateMask(img.select('cs').gte(0.55));
    }).select('B.*').median().clip(region);
  var withCloudPct = s2
    .filter(ee.Filter.lt('CLOUDY_PIXEL_PERCENTAGE', 50))
    .select('B.*').median().clip(region);
  var fallback = ee.Image.constant(0).rename('B8')
    .addBands(ee.Image.constant(0).rename('B4'))
    .addBands(ee.Image.constant(0).rename('B11'))
    .addBands(ee.Image.constant(0).rename('B3'))
    .clip(region);
  return ee.Image(ee.Algorithms.If(
    s2.size().gt(0),
    ee.Image(ee.Algorithms.If(csPlus.size().gt(0), withCs, withCloudPct)),
    fallback
  ));
};

// AlphaEarth 64-dimensional satellite embedding
exports.getEmbeddingImage = function(year, region) {
  var start = ee.Date.fromYMD(year, 1, 1);
  var end   = start.advance(1, 'year');
  return ee.ImageCollection(EMB_COLLECTION)
    .filter(ee.Filter.date(start, end))
    .filter(ee.Filter.bounds(region))
    .mosaic().clip(region);
};

// Annual p10 and p90 percentile composites for phenological indices
exports.getPercentileComposites = function(year, region) {
  var yr     = String(year);
  var csPlus = ee.ImageCollection(CS_COLLECTION)
    .filter(ee.Filter.date(yr+'-01-01', yr+'-12-31'))
    .filter(ee.Filter.bounds(region));
  var masked = ee.ImageCollection(S2_COLLECTION)
    .filter(ee.Filter.date(yr+'-01-01', yr+'-12-31'))
    .filter(ee.Filter.bounds(region))
    .linkCollection(csPlus, csPlus.first().bandNames())
    .map(function(img) {
      return img.updateMask(img.select('cs').gte(0.65));
    }).select('B.*');
  var rename = function(img) {
    return img.rename(
      masked.first().bandNames()
        .map(function(b) { return ee.String(b); })
    ).clip(region);
  };
  return {
    p10: rename(masked.reduce(ee.Reducer.percentile([10]))),
    p90: rename(masked.reduce(ee.Reducer.percentile([90])))
  };
};

File: savana/composites.py
"""Sentinel-2, seasonal, percentile, and AlphaEarth embedding composites.

Ported from ``kogyae.js`` Phase 2 (Sentinel-2 reference imagery) and
Phase 3 (AlphaEarth embeddings), and from the year-loop composite logic
in Phase 8 — the standalone ``composites.js`` module you use in
``mainrun.js`` was empty, so this reconstructs it from the working
monolithic script with the exact same signatures ``mainrun.js`` expects
(``getSentinel2Annual``, ``getSeasonalComposite``, ``getPercentileComposites``,
``getEmbeddingImage``), plus a graceful fallback for seasons/years with
sparse Cloud Score+ coverage.
"""

from __future__ import annotations

from . import config


def sentinel2_annual(year: int, region, cloud_score_threshold: float = 0.65):
    """Cloud-masked median Sentinel-2 SR composite for a calendar year.

    Equivalent to ``getSentinel2Composite`` / ``C.getSentinel2Annual``.
    """
    import ee

    start = ee.Date.fromYMD(year, 1, 1)
    end = start.advance(1, "year")
    s2 = (
        ee.ImageCollection(config.S2_COLLECTION)
        .filter(ee.Filter.date(start, end))
        .filter(ee.Filter.bounds(region))
    )
    cs_plus = ee.ImageCollection(config.CLOUD_SCORE_COLLECTION)
    return (
        s2.linkCollection(cs_plus, cs_plus.first().bandNames())
        .map(lambda img: img.updateMask(img.select("cs").gte(cloud_score_threshold)))
        .select("B.*")
        .median()
        .clip(region)
    )


def seasonal_composite(
    start_date: str,
    end_date: str,
    region,
    cloud_score_threshold: float = 0.55,
    cloud_pct_fallback: float = 50,
):
    """Cloud-masked median Sentinel-2 composite for an arbitrary date range.

    Falls back to a simple ``CLOUDY_PIXEL_PERCENTAGE`` filter if Cloud
    Score+ has no coverage for the window, and to a zero-filled image
    if there are no scenes at all (keeps downstream band math from
    failing on sparse early-record years). Equivalent to
    ``getSeasonalComposite`` / ``C.getSeasonalComposite``.
    """
    import ee

    cs_plus = (
        ee.ImageCollection(config.CLOUD_SCORE_COLLECTION)
        .filter(ee.Filter.date(start_date, end_date))
        .filter(ee.Filter.bounds(region))
    )
    s2 = (
        ee.ImageCollection(config.S2_COLLECTION)
        .filter(ee.Filter.date(start_date, end_date))
        .filter(ee.Filter.bounds(region))
    )
    has_scenes = s2.size().gt(0)
    has_cs = cs_plus.size().gt(0)

    with_cs = (
        s2.linkCollection(cs_plus, cs_plus.first().bandNames())
        .map(lambda img: img.updateMask(img.select("cs").gte(cloud_score_threshold)))
        .select("B.*")
        .median()
        .clip(region)
    )
    with_cloud_pct = (
        s2.filter(ee.Filter.lt("CLOUDY_PIXEL_PERCENTAGE", cloud_pct_fallback))
        .select("B.*")
        .median()
        .clip(region)
    )
    empty = (
        ee.Image.constant(0)
        .rename("B8")
        .addBands(ee.Image.constant(0).rename("B4"))
        .addBands(ee.Image.constant(0).rename("B11"))
        .addBands(ee.Image.constant(0).rename("B3"))
        .clip(region)
    )
    return ee.Image(
        ee.Algorithms.If(
            has_scenes,
            ee.Image(ee.Algorithms.If(has_cs, with_cs, with_cloud_pct)),
            empty,
        )
    )


def percentile_composites(
    year: int,
    region,
    percentiles: tuple[int, int] = (10, 90),
    cloud_score_threshold: float = 0.65,
):
    """Per-band percentile composites (default p10/p90) for a calendar year.

    Used to build the seasonal-amplitude / stability indices.
    Equivalent to ``C.getPercentileComposites`` — returns a dict keyed
    ``"p{percentile}"`` (e.g. ``{"p10": image, "p90": image}``).
    """
    import ee

    start = ee.Date.fromYMD(year, 1, 1)
    end = start.advance(1, "year")
    cs_plus = (
        ee.ImageCollection(config.CLOUD_SCORE_COLLECTION)
        .filter(ee.Filter.date(start, end))
        .filter(ee.Filter.bounds(region))
    )
    s2_masked = (
        ee.ImageCollection(config.S2_COLLECTION)
        .filter(ee.Filter.date(start, end))
        .filter(ee.Filter.bounds(region))
        .linkCollection(cs_plus, cs_plus.first().bandNames())
        .map(lambda img: img.updateMask(img.select("cs").gte(cloud_score_threshold)))
        .select("B.*")
    )
    band_names = s2_masked.first().bandNames()
    out = {}
    for p in percentiles:
        img = (
            s2_masked.reduce(ee.Reducer.percentile([p])).rename(band_names).clip(region)
        )
        out[f"p{p}"] = img
    return out


def embedding_image(year: int, region):
    """Annual AlphaEarth satellite embedding image (64 bands: A01..A64).

    Equivalent to ``getEmbeddingImage`` / ``C.getEmbeddingImage``.
    """
    import ee

    start = ee.Date.fromYMD(year, 1, 1)
    end = start.advance(1, "year")
    return (
        ee.ImageCollection(config.EMBEDDING_COLLECTION)
        .filter(ee.Filter.date(start, end))
        .filter(ee.Filter.bounds(region))
        .mosaic()
        .clip(region)
    )
